flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperHaServicesTest.java

```diff
@@ -0,0 +1,250 @@
+		final Configuration configuration = createConfiguration(flinkPath);
+
+		final TestingBlobStoreService blobStoreService = new TestingBlobStoreService();
+
+		final String unclePath = prefix + "/foobar";
+		client.create().creatingParentContainersIfNeeded().forPath(unclePath);
+
+		runCleanupTest(
+			configuration,
+			blobStoreService,
+			ZooKeeperHaServices::closeAndCleanupAllData);
+
+		assertThat(blobStoreService.isClosedAndCleanedUpAllData(), is(true));
+
+		assertThat(client.checkExists().forPath(flinkPath), is(nullValue()));
+		assertThat(client.checkExists().forPath(unclePath), is(notNullValue()));
+	}
+
+	private static CuratorFramework startCuratorFramework() {
+		return CuratorFrameworkFactory.builder()
+				.connectString(ZOO_KEEPER_RESOURCE.getConnectString())
+				.retryPolicy(new RetryForever(100))
+				.build();
+	}
+
+	@Nonnull
+	private Configuration createConfiguration(String rootPath) {
+		final Configuration configuration = new Configuration();
+		configuration.setString(HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, ZOO_KEEPER_RESOURCE.getConnectString());
+		configuration.setString(HighAvailabilityOptions.HA_ZOOKEEPER_ROOT, rootPath);
+		return configuration;
+	}
+
+	private void runCleanupTest(
+			Configuration configuration,
+			TestingBlobStoreService blobStoreService,
+			ThrowingConsumer<ZooKeeperHaServices, Exception> zooKeeperHaServicesConsumer) throws Exception {
+		try (ZooKeeperHaServices zooKeeperHaServices = new ZooKeeperHaServices(
+			ZooKeeperUtils.startCuratorFramework(configuration),
+			Executors.directExecutor(),
+			configuration,
+			blobStoreService)) {
+
+			// create some Zk services to trigger the generation of paths
+			final LeaderRetrievalService resourceManagerLeaderRetriever = zooKeeperHaServices.getResourceManagerLeaderRetriever();
+			final LeaderElectionService resourceManagerLeaderElectionService = zooKeeperHaServices.getResourceManagerLeaderElectionService();
+			final RunningJobsRegistry runningJobsRegistry = zooKeeperHaServices.getRunningJobsRegistry();
+
+			resourceManagerLeaderRetriever.start(new TestingListener());
+			resourceManagerLeaderElectionService.start(new TestingContender("foobar", resourceManagerLeaderElectionService));
+			final JobID jobId = new JobID();
+			runningJobsRegistry.setJobRunning(jobId);
+
+			resourceManagerLeaderRetriever.stop();
+			resourceManagerLeaderElectionService.stop();
+			runningJobsRegistry.clearJob(jobId);
+
+			zooKeeperHaServicesConsumer.accept(zooKeeperHaServices);
+		}
+	}
+
+	private static class TestingBlobStoreService implements BlobStoreService {
+
+		private boolean closedAndCleanedUpAllData = false;
+		private boolean closed = false;
+
+		@Override
+		public void closeAndCleanupAllData() {
+			closedAndCleanedUpAllData = true;
+		}
+
+		@Override
+		public void close() throws IOException {
+			closed = true;
+		}
+
+		@Override
+		public boolean put(File localFile, JobID jobId, BlobKey blobKey) throws IOException {
```

Comment:
IOException is not thrown, also in `get()`

Reply:
Good point. Will remove it.